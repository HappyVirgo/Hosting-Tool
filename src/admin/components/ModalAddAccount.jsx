import PropTypes from "prop-types";
import React, {useState, useEffect} from "react";
import {Modal} from "react-bootstrap";
import {FaCircleNotch} from "react-icons/fa";

import airbnbLogo from "../img/airbnb-logo.svg";
import vrboLogo from "../img/vrbo-logo.svg";
import {UserConsumer} from "../providers/UserProvider";

import AddAirbnbAccount from "./AddAirbnbAccount";
import AddHomeAwayAccount from "./AddHomeAwayAccount";
import AddAugustAccount from "./AddAugustAccount";

function ModalAddAccount(props) {
    const {show, onHide, hideCloseButtons, credentials, updateUser, user} = props;
    const [channel, setChannel] = useState("");
    const [selectedChannel, setSelectedChannel] = useState("");
    const [showSpinner, setShowSpinner] = useState(false);
    const [errors, setErrors] = useState({});

    useEffect(() => {
        if (show) {
            loadInspectlet();
            loadHotJar();
            // loadSmartLook();
            const {
                credentials: {type}
            } = props;
            let channel = "";
            let selectedChannel = "";
            if (type) {
                channel = type;
                selectedChannel = type;
            }
            setChannel(channel);
            setSelectedChannel(selectedChannel);
        }
    }, [show]);

    function loadInspectlet() {
        // eslint-disable-next-line no-underscore-dangle
        window.__insp = window.__insp || [];
        // eslint-disable-next-line no-undef
        __insp.push(["wid", 1206788301]);
        // eslint-disable-next-line no-undef
        __insp.push(["tagSession", "Airbnb"]);
        // eslint-disable-next-line no-undef
        __insp.push(["identify", user.username]);
        const ldinsp = () => {
            // eslint-disable-next-line no-underscore-dangle
            if (typeof window.__inspld != "undefined") return;
            // eslint-disable-next-line no-underscore-dangle
            window.__inspld = 1;
            const insp = document.createElement("script");
            insp.type = "text/javascript";
            insp.async = true;
            insp.id = "inspsync";
            insp.src = `${
                // eslint-disable-next-line eqeqeq
                document.location.protocol == "https:" ? "https" : "http"
            }://cdn.inspectlet.com/inspectlet.js?wid=1206788301&r=${Math.floor(
                new Date().getTime() / 3600000
            )}`;
            const x = document.getElementsByTagName("script")[0];
            x.parentNode.insertBefore(insp, x);
        };
        setTimeout(ldinsp, 0);
    }

    // eslint-disable-next-line class-methods-use-this
    function loadHotJar() {
        // eslint-disable-next-line func-names
        (function (h, o, t, j, a, r) {
            h.hj =
                h.hj ||
                // eslint-disable-next-line func-names
                function () {
                    // eslint-disable-next-line prefer-rest-params
                    (h.hj.q = h.hj.q || []).push(arguments);
                };
            // eslint-disable-next-line no-underscore-dangle
            h._hjSettings = {hjid: 1683480, hjsv: 6};
            // eslint-disable-next-line prefer-destructuring
            a = o.getElementsByTagName("head")[0];
            r = o.createElement("script");
            r.async = 1;
            // eslint-disable-next-line no-underscore-dangle
            r.src = t + h._hjSettings.hjid + j + h._hjSettings.hjsv;
            a.appendChild(r);
        })(window, document, "https://static.hotjar.com/c/hotjar-", ".js?sv=");
    }

    //  function loadSmartLook() {
    //     window.smartlook ||
    //         (function(d) {
    //             var o = (smartlook = function() {
    //                 o.api.push(arguments);
    //             });
    //             const h = d.getElementsByTagName("head")[0];
    //             const c = d.createElement("script");
    //             o.api = new Array();
    //             c.async = true;
    //             c.type = "text/javascript";
    //             c.charset = "utf-8";
    //             c.src = "https://rec.smartlook.com/recorder.js";
    //             h.appendChild(c);
    //         })(document);
    //     smartlook("init", "44686d86b7f0014741e8b1ae9932d0af1d5f59e6");
    //     smartlook("identify", user.username, {
    //         name: `${user.firstName} ${user.lastName}`,
    //         email: user.username
    //     });
    // }

    function handleToggle(selectedChannel) {
        setSelectedChannel(selectedChannel);
    }

    function selectChannel() {
        if (selectedChannel === "") {
            const newErrors = {};
            newErrors.error = "Please select a channel.";
            setErrors(errors => {
                return {...errors, ...newErrors};
            });
        } else {
            const newErrors = {};
            newErrors.error = "";
            setErrors(errors => {
                return {...errors, ...newErrors};
            });
            setChannel(selectedChannel);
        }
    }

    return (
        <Modal show={show} onHide={onHide} backdrop="static">
            {channel === "Airbnb" && (
                <AddAirbnbAccount
                    onHide={onHide}
                    hideCloseButtons={hideCloseButtons}
                    credentials={credentials}
                    updateUser={updateUser}
                    user={user}
                />
            )}
            {channel === "HomeAway" && (
                <AddHomeAwayAccount
                    onHide={onHide}
                    hideCloseButtons={hideCloseButtons}
                    credentials={credentials}
                    updateUser={updateUser}
                    user={user}
                />
            )}
            {channel === "August" && (
                <AddAugustAccount
                    onHide={onHide}
                    hideCloseButtons={hideCloseButtons}
                    credentials={credentials}
                    updateUser={updateUser}
                    user={user}
                />
            )}
            {channel === "" && (
                <form>
                    <Modal.Header closeButton={!hideCloseButtons}>
                        <Modal.Title>Link Account</Modal.Title>
                    </Modal.Header>

                    <Modal.Body className="pd-20 pd-sm-40">
                        <p>
                            Host Tools links with channels like Airbnb and VRBO to automate the
                            management of your listing.
                        </p>
                        <p>Please select a channel you would like to link with Host Tools.</p>
                        <div className="form-group mg-t-50">
                            <div className="row">
                                <div className="col-4 d-flex justify-content-end align-items-center">
                                    <label className="ckbox">
                                        <input
                                            id="selectedChannelAirbnb"
                                            data-testid="airbnb-checkbox"
                                            type="checkbox"
                                            checked={selectedChannel === "Airbnb"}
                                            onChange={() => {
                                                handleToggle("Airbnb");
                                            }}
                                        />
                                        <span />
                                    </label>
                                </div>
                                <div
                                    data-testid="aribnb-logo-wrapper"
                                    className="col-8 d-flex justify-content-start align-items-center c-pointer"
                                    onClick={() => {
                                        handleToggle("Airbnb");
                                    }}
                                    onKeyPress={() => {
                                        handleToggle("Airbnb");
                                    }}
                                    role="presentation"
                                >
                                    <img src={airbnbLogo} alt="Airbnb" height="35" />
                                </div>
                            </div>
                        </div>
                        <div className="form-group">
                            <div className="row">
                                <div className="col-4 d-flex justify-content-end align-items-center">
                                    <label className="ckbox">
                                        <input
                                            id="selectedChannelHomeAway"
                                            data-testid="homeaway-checkbox"
                                            type="checkbox"
                                            checked={selectedChannel === "HomeAway"}
                                            onChange={() => {
                                                handleToggle("HomeAway");
                                            }}
                                        />
                                        <span />
                                    </label>
                                </div>
                                <div
                                    data-testid="homeaway-logo-wrapper"
                                    className="col-8 d-flex justify-content-start align-items-center c-pointer"
                                    onClick={() => {
                                        handleToggle("HomeAway");
                                    }}
                                    onKeyPress={() => {
                                        handleToggle("HomeAway");
                                    }}
                                    role="presentation"
                                >
                                    <img src={vrboLogo} alt="VRBO" height="35" />
                                </div>
                            </div>
                        </div>
                        {errors.error && <div className="alert alert-danger">{errors.error}</div>}
                    </Modal.Body>

                    <Modal.Footer>
                        {!hideCloseButtons && (
                            <button type="button" className="btn btn-outline-dark" onClick={onHide}>
                                Cancel
                            </button>
                        )}
                        <button
                            data-testid="btn-next"
                            type="button"
                            className="btn btn-outline-primary"
                            onClick={selectChannel}
                        >
                            {showSpinner && <FaCircleNotch className="fa-spin mr-1" />}
                            Next
                        </button>
                    </Modal.Footer>
                </form>
            )}
        </Modal>
    );
}

ModalAddAccount.propTypes = {
    show: PropTypes.bool.isRequired,
    onHide: PropTypes.func.isRequired,
    hideCloseButtons: PropTypes.bool,
    credentials: PropTypes.shape({
        type: PropTypes.string,
        airbnbUsername: PropTypes.string,
        airbnbPassword: PropTypes.string
    }),
    updateUser: PropTypes.func.isRequired,
    user: PropTypes.shape({
        username: PropTypes.string,
        firstName: PropTypes.string,
        lastName: PropTypes.string
    }).isRequired
};

ModalAddAccount.defaultProps = {
    credentials: {
        type: "",
        airbnbUsername: "",
        airbnbPassword: ""
    },
    hideCloseButtons: false
};

const ConnectedModalAddAccount = props => (
    <UserConsumer>
        {({user, updateUser}) => <ModalAddAccount {...props} user={user} updateUser={updateUser} />}
    </UserConsumer>
);
export default ConnectedModalAddAccount;
