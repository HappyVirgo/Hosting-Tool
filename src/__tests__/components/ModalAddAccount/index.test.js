import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";

import ModalAddAccount from "@/admin/components/ModalAddAccount";

jest.mock("@/admin/providers/UserProvider", () => {
  const UserContext = jest.requireActual("react").createContext({
    user: {},
    updateUser: jest.fn(),
  });

  return {
    __esModule: true,
    UserContext,
    UserConsumer: UserContext.Consumer,
  };
});

const setup = (overrides) => {
  const props = {
    onHide: jest.fn(),

    ...overrides,
  };

  const wrapper = render(<ModalAddAccount {...props} />);

  return {
    wrapper,
    props,
  };
};

describe("ModalAddAccount", () => {
  ["Airbnb", "HomeAway", "August"].forEach((type) => {
    test(`should show modal link ${type}`, () => {
      jest.useFakeTimers();

      const { wrapper } = setup({
        show: true,
        credentials: {
          type,
        },
      });

      jest.runAllTimers();

      expect(wrapper.baseElement).toMatchSnapshot();
    });
  });

  test("should show error without selecting a channel", () => {
    const { wrapper } = setup({
      show: true,
      credentials: {
        type: "",
      },
    });

    fireEvent.click(wrapper.queryByTestId("btn-next"));

    expect(wrapper.queryByText("Please select a channel.")).toBeInTheDocument();
  });

  [
    ["airbnb-checkbox", "Link Airbnb Account"],
    ["aribnb-logo-wrapper", "Link Airbnb Account"],
    ["homeaway-checkbox", "Link VRBO Account"],
    ["homeaway-logo-wrapper", "Link VRBO Account"],
  ].forEach(([selector, result]) => {
    test(`should display ${result} as click on ${
      /checkbox$/.test(selector) ? "checkbox" : "wrapper"
    }`, () => {
      const { wrapper } = setup({
        show: true,
        credentials: {
          type: "",
        },
      });

      fireEvent.click(screen.queryByTestId(selector));
      fireEvent.click(wrapper.queryByTestId("btn-next"));

      expect(wrapper.queryByText(result)).toBeInTheDocument();
    });
  });
});
