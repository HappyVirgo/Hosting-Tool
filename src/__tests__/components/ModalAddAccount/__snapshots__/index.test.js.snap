// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalAddAccount should show modal link Airbnb 1`] = `
<body
  class="modal-open"
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="fade modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="fade modal show"
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog"
      role="document"
    >
      <div
        class="modal-content"
      >
        <form>
          <div
            class="modal-header"
          >
            <div
              class="modal-title h4"
            >
              Link Airbnb Account
            </div>
            <button
              class="close"
              type="button"
            >
              <span
                aria-hidden="true"
              >
                ×
              </span>
              <span
                class="sr-only"
              >
                Close
              </span>
            </button>
          </div>
          <div
            class="pd-20 pd-sm-40 modal-body"
          >
            <p>
              Host Tools will request an authentication token that will be used to automatically send messages to your guests or set your listing's prices on your behalf. It will only send messages you tell it to send and will never make any changes to your account without your permission.
            </p>
            <p>
              All network traffic is encrypted on Host Tools and all data is stored using the latest compliance standards.
            </p>
            <div
              class="pd-md-20"
            >
              <div
                class="mg-b-20 d-flex justify-content-start align-items-center"
              >
                <img
                  alt="Airbnb"
                  class="rounded-circle mg-r-20"
                  height="65"
                  src="airbnb-logo-solid-square.svg"
                />
                <div>
                  <h2
                    class="mg-b-0"
                  >
                    Airbnb Account
                  </h2>
                  <span
                    class="text-muted text-uppercase tx-11"
                  >
                    Link your account with Airbnb credentials
                  </span>
                </div>
              </div>
              <div
                class="form-group"
              >
                <label
                  class="az-content-label tx-11 tx-medium tx-gray-600"
                  for="airbnbUsername"
                >
                  Airbnb Email
                  <svg
                    class="text-muted ml-1"
                    fill="none"
                    height="1em"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <circle
                      cx="12"
                      cy="12"
                      r="10"
                    />
                    <path
                      d="M9.09 9a3 3 0 0 1 5.83 1c0 2-3 3-3 3"
                    />
                    <line
                      x1="12"
                      x2="12"
                      y1="17"
                      y2="17"
                    />
                  </svg>
                </label>
                <input
                  autocomplete="off"
                  class="form-control"
                  id="airbnbUsername"
                  name="airbnbUsername"
                  placeholder="Airbnb Email..."
                  required=""
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-group"
              >
                <label
                  class="az-content-label tx-11 tx-medium tx-gray-600"
                  for="airbnbPassword"
                >
                  Airbnb Password
                  <svg
                    class="text-muted ml-1"
                    fill="none"
                    height="1em"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <circle
                      cx="12"
                      cy="12"
                      r="10"
                    />
                    <path
                      d="M9.09 9a3 3 0 0 1 5.83 1c0 2-3 3-3 3"
                    />
                    <line
                      x1="12"
                      x2="12"
                      y1="17"
                      y2="17"
                    />
                  </svg>
                </label>
                <input
                  autocomplete="off"
                  class="form-control"
                  id="airbnbPassword"
                  name="airbnbPassword"
                  placeholder="Airbnb Password..."
                  required=""
                  type="password"
                  value=""
                />
              </div>
            </div>
            <div
              class="alert alert-info"
            >
              Do you use Google or Facebook to log into your Airbnb account? To connect an Airbnb property with Host Tools, you'll need to first 
              <a
                href="https://www.airbnb.com/help/article/76/how-do-i-reset-my-password"
                rel="noopener noreferrer"
                target="_blank"
              >
                create a password
              </a>
               for your Airbnb account.
            </div>
          </div>
          <div
            class="modal-footer"
          >
            <button
              class="btn btn-outline-dark"
              type="button"
            >
              Cancel
            </button>
            <button
              class="btn btn-outline-primary"
              type="button"
            >
              Add Account
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</body>
`;

exports[`ModalAddAccount should show modal link August 1`] = `
<body
  class="modal-open"
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="fade modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="fade modal show"
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog"
      role="document"
    >
      <div
        class="modal-content"
      >
        <form>
          <div
            class="modal-header"
          >
            <div
              class="modal-title h4"
            >
              Link August Account
            </div>
            <button
              class="close"
              type="button"
            >
              <span
                aria-hidden="true"
              >
                ×
              </span>
              <span
                class="sr-only"
              >
                Close
              </span>
            </button>
          </div>
          <div
            class="pd-20 pd-sm-40 modal-body"
          >
            <p>
              Host Tools will request an authentication token that will be used to automatically send messages to your guests or set your listing's prices on your behalf. It will only send messages you tell it to send and will never make any changes to your account without your permission.
            </p>
            <p>
              All network traffic is encrypted on Host Tools and all data is stored using the latest compliance standards.
            </p>
            <div
              class="pd-md-20"
            >
              <div
                class="mg-b-20 d-flex justify-content-start align-items-center"
              >
                <img
                  alt="August Smart Locks"
                  class="rounded-circle mg-r-20"
                  height="65"
                  src="august-logo-solid-square.svg"
                />
                <div>
                  <h2
                    class="mg-b-0"
                  >
                    August Account
                  </h2>
                  <span
                    class="text-muted text-uppercase tx-11"
                  >
                    Link your account with August credentials
                  </span>
                </div>
              </div>
              <div
                class="form-group"
              >
                <label
                  class="az-content-label tx-11 tx-medium tx-gray-600"
                  for="username"
                >
                  August Email
                  <svg
                    class="text-muted ml-1"
                    fill="none"
                    height="1em"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <circle
                      cx="12"
                      cy="12"
                      r="10"
                    />
                    <path
                      d="M9.09 9a3 3 0 0 1 5.83 1c0 2-3 3-3 3"
                    />
                    <line
                      x1="12"
                      x2="12"
                      y1="17"
                      y2="17"
                    />
                  </svg>
                </label>
                <input
                  autocomplete="off"
                  class="form-control"
                  id="username"
                  name="username"
                  placeholder="August Email..."
                  required=""
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-group"
              >
                <label
                  class="az-content-label tx-11 tx-medium tx-gray-600"
                  for="password"
                >
                  August Password
                  <svg
                    class="text-muted ml-1"
                    fill="none"
                    height="1em"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <circle
                      cx="12"
                      cy="12"
                      r="10"
                    />
                    <path
                      d="M9.09 9a3 3 0 0 1 5.83 1c0 2-3 3-3 3"
                    />
                    <line
                      x1="12"
                      x2="12"
                      y1="17"
                      y2="17"
                    />
                  </svg>
                </label>
                <input
                  autocomplete="off"
                  class="form-control"
                  id="password"
                  name="password"
                  placeholder="August Password..."
                  required=""
                  type="password"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="modal-footer"
          >
            <button
              class="btn btn-outline-dark"
              type="button"
            >
              Cancel
            </button>
            <button
              class="btn btn-outline-primary"
              type="button"
            >
              Add Account
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</body>
`;

exports[`ModalAddAccount should show modal link HomeAway 1`] = `
<body
  class="modal-open"
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="fade modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="fade modal show"
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog"
      role="document"
    >
      <div
        class="modal-content"
      >
        <form>
          <div
            class="modal-header"
          >
            <div
              class="modal-title h4"
            >
              Link VRBO Account
            </div>
            <button
              class="close"
              type="button"
            >
              <span
                aria-hidden="true"
              >
                ×
              </span>
              <span
                class="sr-only"
              >
                Close
              </span>
            </button>
          </div>
          <div
            class="pd-20 pd-sm-40 modal-body"
          >
            <p>
              Host Tools will request an authentication token that will be used to automatically send messages to your guests or set your listing's prices on your behalf. It will only send messages you tell it to send and will never make any changes to your account without your permission.
            </p>
            <p>
              All network traffic is encrypted on Host Tools and all data is stored using the latest compliance standards.
            </p>
            <div
              class="pd-md-20"
            >
              <div
                class="mg-b-20 d-flex justify-content-start align-items-center"
              >
                <img
                  alt="VRBO"
                  class="rounded-circle mg-r-20"
                  height="65"
                  src="vrbo-logo-solid-square.svg"
                />
                <div>
                  <h2
                    class="mg-b-0"
                  >
                    VRBO Account
                  </h2>
                  <span
                    class="text-muted text-uppercase tx-11"
                  >
                    Link your account with VRBO credentials
                  </span>
                </div>
              </div>
              <div
                class="form-group"
              >
                <label
                  class="az-content-label tx-11 tx-medium tx-gray-600"
                  for="username"
                >
                  VRBO Email
                  <svg
                    class="text-muted ml-1"
                    fill="none"
                    height="1em"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <circle
                      cx="12"
                      cy="12"
                      r="10"
                    />
                    <path
                      d="M9.09 9a3 3 0 0 1 5.83 1c0 2-3 3-3 3"
                    />
                    <line
                      x1="12"
                      x2="12"
                      y1="17"
                      y2="17"
                    />
                  </svg>
                </label>
                <input
                  autocomplete="off"
                  class="form-control"
                  id="username"
                  name="username"
                  placeholder="VRBO Email..."
                  required=""
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-group"
              >
                <label
                  class="az-content-label tx-11 tx-medium tx-gray-600"
                  for="password"
                >
                  VRBO Password
                  <svg
                    class="text-muted ml-1"
                    fill="none"
                    height="1em"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <circle
                      cx="12"
                      cy="12"
                      r="10"
                    />
                    <path
                      d="M9.09 9a3 3 0 0 1 5.83 1c0 2-3 3-3 3"
                    />
                    <line
                      x1="12"
                      x2="12"
                      y1="17"
                      y2="17"
                    />
                  </svg>
                </label>
                <input
                  autocomplete="off"
                  class="form-control"
                  id="password"
                  name="password"
                  placeholder="VRBO Password..."
                  required=""
                  type="password"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="modal-footer"
          >
            <button
              class="btn btn-outline-dark"
              type="button"
            >
              Cancel
            </button>
            <button
              class="btn btn-outline-primary"
              type="button"
            >
              Add Account
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</body>
`;
